### PR TITLE
metrics endpoint in pepr

### PR DESCRIPTION
**Pull request by bdw617**
added a simple metrics route to pepr, not exposed via servicemonitor a WIP for now.

for now, I've only added 3 counters that live in the /mutate endpoint:
alert: if you use the wrong `API KEY`
summary: it uses a full summary for OK requests. 
error: a counter if there are issues. 

I've not yet figured out the best way to allow extended metrics into each of the capabilities, but for now, the simplicity of this should help us drive that.

https://github.com/defenseunicorns/pepr/issues/24

### CODE REVIEW

File: src/lib/controller.ts [hunk 1]
```diff
@@ -121,14 +127,26 @@ export class Controller {
...
+    try {
+      res.send(await this.metricsCollector.getMetrics());
+    } catch (err) {
+      console.error(err);
+      res.status(500).send("Internal Server Error");
+    }
+  };
+
   private mutate = async (req: express.Request, res: express.Response) => {
+    const startTime = performance.now();
```
Comment (jeff-mccoy) [inline src/lib/controller.ts]: would probably be better to just use the request time tracking in the `logger` function

File: src/lib/metrics.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    });
+
+    this.summary = new promClient.Summary({
+      name: `${prefix}_summary`,
+      help: "summary",
+      registers: [this.registery],
+    });
+  }
+
+  error(): void {
```
Comment (jeff-mccoy) [inline src/lib/metrics.ts]: We should at least have jsdoc for public functions (`/** */` format).

File: src/lib/metrics.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    this.errors.inc();
+  }
+
+  alert(): void {
+    this.alerts.inc();
+  }
+
+  observe(startTime: number): void {
```
Comment (jeff-mccoy) [inline src/lib/metrics.ts]: it feels odd to me to have a time entry pulled by the callee to start and then also in the called method to end--seems like it should be generated and tracked internally, if nothing else:

E.g. 

```typescript
observe() {
  this._timer = performance.now();
}

observeEnd() {
this.summary.observe(performance.now() - this._timer);
}
```

File: src/lib/metrics.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    this.summary.observe(performance.now() - startTime);
+  }
+
+  async getMetrics(): Promise<string> {
```
Comment (jeff-mccoy) [inline src/lib/metrics.ts]: style-wise, we don't typically include the return info unless it is needed; it just makes the code more verbose and doesn't help intellisense for the most part.

File: src/lib/metrics.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023-Present The Pepr Authors
+
+import promClient from "prom-client";
+import { performance } from "perf_hooks";
+
+export class MetricsCollector {
+  private registery: promClient.Registry;
```
Comment (jeff-mccoy) [inline src/lib/metrics.ts]: spelling `registery`

File: src/lib/metrics.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  private errors: promClient.Counter<string>;
```
Comment (jeff-mccoy) [inline src/lib/metrics.ts]: ironically, everywhere but the `controller.ts`, we are using the same convention of prefixing the private fields with an `_`,:

`private _errors: promClient.Counter<string>;`

**Review by jeff-mccoy — CHANGES_REQUESTED** [review]
This will be a very useful addition, good work. More code comments would be nice--especially jsdoc for public methods, we need to add some test to the e2e suite to verify the metrics are working. Also, could you drop a sample metrics output in the PR comments for reference, assuming it will output [Prometheus format](https://prometheus.io/docs/instrumenting/exposition_formats/)?

File: src/lib/metrics.ts (hunk 3 above)
Comment (bdw617) [reply src/lib/metrics.ts]: I'd have to create a new instance of the class for that...

How's this instead?
  observeStart(): number {
    return performance.now()
  }

  observeEnd(startTime: number): void {
    this.summary.observe(performance.now() - startTime);
  }

File: src/lib/metrics.ts (hunk 4 above)
Comment (bdw617) [reply src/lib/metrics.ts]: :| 

so like this instead?

  async getMetrics() {
    return this.registry.metrics();
  }

Comment (jeff-mccoy) [reply src/lib/metrics.ts]: yes plz!

File: src/lib/metrics.ts (hunk 3 above)
Comment (jeff-mccoy) [reply src/lib/metrics.ts]: yeah that's prob fine, thx

**Comment by jeff-mccoy** [comment]
@bdw617 still need a test for this to get it merged.

**Review by jeff-mccoy — APPROVED** [review]
lgtm, thanks!

**Pull request merged**